robot/impl/local_robot_test.go

```diff
@@ -3359,3 +3360,174 @@ func TestCloudMetadata(t *testing.T) {
 		})
 	})
 }
+
+func TestReconfigureOnModuleRename(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	// Precompile complex module to avoid timeout issues when building takes too long.
+	complexPath := rtestutils.BuildTempModule(t, "examples/customresources/demos/complexmodule")
+
+	// Manually define mybase model, as importing it can cause double registration.
+	myBaseModel := resource.NewModel("acme", "demo", "mybase")
+
+	// Create config with at least one module and a component from that module
+	cfg := &config.Config{
+		Modules: []config.Module{
+			{
+				Name:     "mod",
+				ExePath:  complexPath,
+				LogLevel: "info",
+			},
+		},
+		Components: []resource.Config{
+			{
+				Name:  "myBase",
+				API:   base.API,
+				Model: myBaseModel,
+				Attributes: rutils.AttributeMap{
+					"motorL": "motor1",
+					"motorR": "motor2",
+				},
+			},
+			{
+				Name:                "motor1",
+				API:                 motor.API,
+				Model:               fakeModel,
+				ConvertedAttributes: &fakemotor.Config{},
+			},
+			{
+				Name:                "motor2",
+				API:                 motor.API,
+				Model:               fakeModel,
+				ConvertedAttributes: &fakemotor.Config{},
+			},
+		},
+	}
+
+	// Create copy of cfg since Reconfigure (called when setting up a robot) modifies cfg.
+	cfgCopy := *cfg
+	r := setupLocalRobot(t, ctx, cfg, logger)
+
+	// Use copy to modify config by renaming module
+	cfgCopy.Modules[0].Name = "mod-renamed"
+
+	// Create copy of cfgCopy to test against
+	expectedCfg := cfgCopy
+	r.Reconfigure(ctx, &cfgCopy)
+
+	actualCfg := r.Config()
+
+	// Verify attributes of config
+
+	// Manually inspect components
+	test.That(t, len(actualCfg.Components), test.ShouldEqual, 3)
+	for _, comp := range actualCfg.Components {
+		isMyBase := comp.Equals(cfg.Components[0])
+		isMotor1 := comp.Equals(cfg.Components[1])
+		isMotor2 := comp.Equals(cfg.Components[2])
+		test.That(t, isMyBase || isMotor1 || isMotor2, test.ShouldBeTrue)
+	}
+
+	// Manually inspect module
+	test.That(t, len(actualCfg.Modules), test.ShouldEqual, 1)
+	test.That(t, actualCfg.Modules[0].Equals(expectedCfg.Modules[0]), test.ShouldBeTrue)
```

Comment:
what do these assert statements actually test? Would these have failed before your patch?

Maybe a good one would be to check ResourceNames and see that myBase is still there

Reply:
clarification/follow-up on what I said irl: old components from newly added (read: renamed) modules would exist in e.g. `conf.Right.components` but weren't modified or truly "added" so wouldn't show up in `conf.Components.Modified` or `conf.Components.Added`, so this test would fail because they would literally not end up in the updated config (and from there get built). I agree though that the test would be more robust to check that the component actually gets built as well

Reply:
these assertions aren't testing `conf.Components.Modified` or `conf.Components.Added` tho (which would be valid to test if we cared about intermediate states). It is testing that actualCfg is being constructed properly based on the resource graph, since that's what `r.Config()` does.

Based on that, I think a check for the correct resources is sufficient and more correct than getting the config from the robot to assert on

Reply:
oh I see, I flipped logic (i.e. the resource graph is built from `Config()`) but if `Config()` is built from the resource graph then it definitely makes more sense / is more correct to test that resources get added to the resource graph by querying for them directly. will change 🫡